247 Promote deployment

martyspiewak opened the pull request:
Closes #247

waciumawanjohi committed:
Promote deployment

- If object stamped out by ClusterDeploymentTemplate passes the provided
  conditions, objects later in the delivery chain can consume the source
  that was passed into deployment.

Co-authored-by: Marty Spiewak <mspiewak@vmware.com>

cirocosta commented on pkg/apis/v1alpha1/cluster_deployment_template.go:
```diff
@@ -41,6 +66,41 @@ type ClusterDeploymentTemplateList struct {
+	return nil
+}
+
+func bothSet(spec DeploymentSpec) bool {
```
`bothSet` and `neitherSet` here end up being available (although privately) for the whole v1alpha1 package. given that _it is_ DeploymentSpec specific (given the signature of the functions), I wonder if it'd be better to instead make them DeploymentSpec methods instead 

```suggestion
func (spec DeploymentSpec) both_somethingsomething_Set() bool {
```

wdyt?

waciumawanjohi replied:
Good call

cirocosta commented on pkg/templates/cluster_deployment_template.go:
```diff
@@ -37,14 +39,110 @@ func (t clusterDeploymentTemplate) GetName() string {
-	return &Output{}, nil
+func (t clusterDeploymentTemplate) GetOutput(stampedObject *unstructured.Unstructured, templatingContext map[string]interface{}) (*Output, error) {
+	if err := t.outputReady(stampedObject); err != nil {
+		return nil, err
```
not sure if intended or not - should this err be wrapped?

waciumawanjohi replied:
Better to leave it unwrapped. Carto throws typed errors if the output is not ready (e.g. the conditions on the Deployment are not met) and wrapping means that our later error handling must unwrap them further.

waciumawanjohi committed:
Repository get templates returns apiTemplates

Repository package should not have responsibility for converting
objects from api objects

waciumawanjohi committed:
Repository does not use template package

waciumawanjohi committed:
GetOutput takes no parameters

Templates have different requirements for fulfilling
GetOutput. Ensure that these are set before calling.

waciumawanjohi committed:
Make functions private to DeploymentTemplate

emmjohnson commented on pkg/controller/deliverable/reconciler.go:
```diff
@@ -90,7 +91,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 			r.conditionManager.AddPositive(TemplateRejectedByAPIServerCondition(typedErr))
 		case realizer.RetrieveOutputError:
-			r.conditionManager.AddPositive(MissingValueAtPathCondition(typedErr.ResourceName(), typedErr.JsonPathExpression()))
+			switch typedErr.Err.(type) {
```
I'm not sold on this, but I think we can work on this outside of this PR (my next PR for handing errors maybe?)

martyspiewak replied:
Sounds good
